From c5241e94030f774df5f410e07c4de894d41e64e2 Mon Sep 17 00:00:00 2001
From: Rogerio Guerra Borin <rogerio.borin@toradex.com>
Date: Tue, 6 Feb 2024 23:14:20 -0300
Subject: [PATCH 1/2] mount: Allow building when macro MOUNT_ATTR_IDMAP is not
 available

This is to allow building the software on machines not having the
MOUNT_ATTR_IDMAP macro in header "linux/mount.h". When that macro is not
available, the dependency on struct mount_attr is also eliminated (which
is good since both the macro and the struct were added to the kernel
uapi virtually at the same time).

With the changes in this commit, errors would be thrown at runtime when
mounting the erofs image, but only if the idmap feature is used; this
resembles the behavior when the "new mount API" is not detected.

Upstream-Status: Accepted [https://github.com/containers/composefs/pull/253]

Signed-off-by: Rogerio Guerra Borin <rogerio.borin@toradex.com>
---
 configure.ac              | 11 +++++++++++
 libcomposefs/lcfs-mount.c |  6 ++++++
 2 files changed, 17 insertions(+)

diff --git a/configure.ac b/configure.ac
index b8ff154..1aa49e4 100644
--- a/configure.ac
+++ b/configure.ac
@@ -16,6 +16,17 @@ m4_ifdef([PKG_INSTALLDIR], [PKG_INSTALLDIR], AC_SUBST([pkgconfigdir], ${libdir}/
 PKGCONFIG_REQUIRES=
 PKGCONFIG_REQUIRES_PRIVATELY=
 
+AC_MSG_CHECKING([for MOUNT_ATTR_IDMAP])
+AC_COMPILE_IFELSE(
+	[AC_LANG_PROGRAM([
+			#include <sys/mount.h>
+			#include <linux/mount.h>
+	],[int foo = MOUNT_ATTR_IDMAP;]
+	)],
+	[AC_MSG_RESULT(yes)
+		AC_DEFINE([HAVE_MOUNT_ATTR_IDMAP], 1, [Define if MOUNT_ATTR_IDMAP is available in linux/mount.h])],
+	[AC_MSG_RESULT(no)])
+
 AC_MSG_CHECKING([for new mount API (fsconfig)])
 AC_COMPILE_IFELSE(
 	[AC_LANG_SOURCE([[
diff --git a/libcomposefs/lcfs-mount.c b/libcomposefs/lcfs-mount.c
index 0a4b08f..5285833 100644
--- a/libcomposefs/lcfs-mount.c
+++ b/libcomposefs/lcfs-mount.c
@@ -108,6 +108,7 @@ static int syscall_move_mount(int from_dfd, const char *from_pathname, int to_df
 #endif
 }
 
+#ifdef HAVE_MOUNT_ATTR_IDMAP
 static int syscall_mount_setattr(int dfd, const char *path, unsigned int flags,
 				 struct mount_attr *attr, size_t usize)
 {
@@ -122,6 +123,7 @@ static int syscall_mount_setattr(int dfd, const char *path, unsigned int flags,
 	return -1;
 #endif
 }
+#endif
 
 #define MAX_DIGEST_SIZE 64
 
@@ -381,6 +383,7 @@ static int lcfs_mount_erofs(const char *source, const char *target,
 		return -errno;
 
 	if (use_idmap) {
+#ifdef HAVE_MOUNT_ATTR_IDMAP
 		struct mount_attr attr = {
 			.attr_set = MOUNT_ATTR_IDMAP,
 			.userns_fd = state->options->idmap_fd,
@@ -390,6 +393,9 @@ static int lcfs_mount_erofs(const char *source, const char *target,
 					    sizeof(struct mount_attr));
 		if (res < 0)
 			return -errno;
+#else
+		return -ENOTSUP;
+#endif
 	}
 
 	res = syscall_move_mount(fd_mnt, "", AT_FDCWD, target,
-- 
2.25.1

